src/uct/cuda/cuda_ipc/cuda_ipc_md.c

```diff
@@ -48,24 +48,91 @@ static ucs_status_t
 uct_cuda_ipc_mem_add_reg(void *addr, uct_cuda_ipc_memh_t *memh,
                          uct_cuda_ipc_lkey_t **key_p)
 {
+    CUipcMemHandle *legacy_handle;
     uct_cuda_ipc_lkey_t *key;
     ucs_status_t status;
+#if HAVE_CUDA_FABRIC
+    CUmemGenericAllocationHandle handle;
+    CUmemoryPool mempool;
+    int legacy_capable;
+#endif
 
-    key = ucs_malloc(sizeof(*key), "uct_cuda_ipc_lkey_t");
+    key = ucs_calloc(1, sizeof(*key), "uct_cuda_ipc_lkey_t");
     if (key == NULL) {
         return UCS_ERR_NO_MEMORY;
     }
 
-    status = UCT_CUDADRV_FUNC(cuIpcGetMemHandle(&key->ph, (CUdeviceptr)addr),
+    legacy_handle = (CUipcMemHandle*)&key->ph;
+    UCT_CUDADRV_FUNC_LOG_ERR(cuMemGetAddressRange(&key->d_bptr, &key->b_len,
+                (CUdeviceptr)addr));
+
+#if HAVE_CUDA_FABRIC
+    /* cuda_ipc can handle VMM, mallocasync, and legacy pinned device so need to
+     * pack appropriate handle */
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuPointerGetAttribute(&legacy_capable,
+                CU_POINTER_ATTRIBUTE_IS_LEGACY_CUDA_IPC_CAPABLE,
+                (CUdeviceptr)addr));
+    if (legacy_capable) {
+        key->ph.handle_type = UCT_CUDA_IPC_KEY_HANDLE_TYPE_LEGACY;
+        legacy_handle       = &key->ph.handle.legacy;
+        goto legacy_path;
+    }
+
+    status =
+        UCT_CUDADRV_FUNC(cuMemRetainAllocationHandle(&handle, addr),
+                UCS_LOG_LEVEL_DIAG);
+    if (status == UCS_OK) {
+        status =
+            UCT_CUDADRV_FUNC_LOG_ERR(cuMemExportToShareableHandle(
+                        &key->ph.handle.fabric_handle, handle,
+                        CU_MEM_HANDLE_TYPE_FABRIC, 0));
+        if (status != UCS_OK) {
+            cuMemRelease(handle);
+            goto err;
+        }
+
+        status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemRelease(handle));
+        if (status != UCS_OK) {
+            goto err;
+        }
+
+        key->ph.handle_type = UCT_CUDA_IPC_KEY_HANDLE_TYPE_VMM;
+        ucs_trace("packed vmm fabric handle for %p", addr);
+        goto common_path;
+    } else {
```

Comment:
so we assume that if cuMemRetainAllocationHandle failed, it's a mempool?
can we check it explicitly by some poster query or as part of cuPointerGetAttribute?\
in any case need to add documentation in the code that describes the logic (in high level)

Reply:
We conclude mempool if it's not legacy ipc memory (viz cudaMalloc), and it is not cuMemcreate/VMM memory (because retainAllocHandle failed).